compiler/vmdef.nim

```diff
@@ -12,17 +12,43 @@
 
 import ast, idents, options, modulegraphs, lineinfos
 
+type TInstrType* = uint64
+
 const
+  regOBits = 8 # Opcode
+  regABits = 8
+  regBBits = 8
+  regCBits = 8
+  regBxBits = 32
+
   byteExcess* = 128 # we use excess-K for immediates
-  wordExcess* = 32768
 
   MaxLoopIterations* = 10_000_000 # max iterations of all loops
 
+const
+
+  # Calculate register shifts, masks and ranges depending on theirs sizes
+
+  regOShift* = 0
+  regAShift* = (regOShift + regOBits).TInstrType
+  regBShift* = (regAShift + regABits).TInstrType
+  regCShift* = (regBShift + regBBits).TInstrType
+  regBxShift* = (regAShift + regABits).TInstrType
+
+  regOMask* = ((1 shl regOBits) - 1)
+  regAMask* = ((1 shl regABits) - 1)
+  regBMask* = ((1 shl regBBits) - 1)
+  regCMask* = ((1 shl regCBits) - 1)
+  regBxMask* = ((1 shl regBxBits) - 1)
+
+  wordExcess* = 1 shl (regBxBits-1)
+  regBxMin* = -wordExcess+1
+  regBxMax* =  wordExcess-1
 
 type
   TRegister* = range[0..255]
```

Comment:
This is also a significant limitation. My project requires more registers.
I propose we increase high(TRegister) and high(TDest) to 65535 or something.

Reply:
@sealmove The bytecode has only 8 bits for the register. So this value isn't something that can be easily increased. But not too long ago I worked with Araq on a better VM register reuse. Maybe all you need is to use the development version of Nim. It surely helped me with the ``too many registers required`` problem.

Reply:
To be clear: increasing A,B and C to more then 8 bits does already seem to solve this problem for us.

And of course, you can also generate different code and find all kind of workarounds for VM limitations, but would life not be easier if this just worked?

Reply:
@zevv Well I don't really like that type of reasoning. With the same type of reasoning you can justify any ignorance towards how the hardware or the operating system works. Sometimes it is better to reflect the limitations of the underlying system is a transparent way than to try finding workarounds to make every badly designed input just work. It is better if you argue in advantages and disadvantages and be honest about them, because there is no "fix" without a penalty. Everything has a cost. In this case, clearly it would be nice if code that uses lots of variables would just work. That is my goal as well. But the solution doesn't always have to be the first thing that comes into mind. 256 registers is in fact a lot of registers. When they go out, it usually means that they are wasted somewhere excessively. Increasing the amount of available registers doesn't fix the problem of the register waste. Maybe it is better to see this problem early instead of late, when the CPU starts to swap memory (exaggeration).

Reply:
Well, I pass my grammar to npeg and I get this error, so it's out of my hands. I am not generating the code, npeg does. Nevertheless I think relaxing the "only 2^8 registers" restriction (to 2^16 registers) does not come with such a significant cost, so I think your argument krux02 is very debatable. Also a solution is to provide the user with a way to tweak the register sizes (through cli argument or cfg). Then we can have the defaults as they are now, but give the allow users with special needs (don't judge what this means for me) to fiddle with the VM and adjust it to their needs.

Reply:
@zevv I should be sorry, those words were a bit harsh. I think all I want is to see the problem on my own before I agree to increase the VM register count.